pkg/resource/table/hooks_global_secondary_indexes.go

```diff
@@ -0,0 +1,274 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package table
+
+import (
+	"context"
+
+	ackcompare "github.com/aws-controllers-k8s/runtime/pkg/compare"
+	ackrtlog "github.com/aws-controllers-k8s/runtime/pkg/runtime/log"
+	ackutil "github.com/aws-controllers-k8s/runtime/pkg/util"
+	"github.com/aws/aws-sdk-go/aws"
+	svcsdk "github.com/aws/aws-sdk-go/service/dynamodb"
+
+	"github.com/aws-controllers-k8s/dynamodb-controller/apis/v1alpha1"
+)
+
+// canUpdateTableGSIs return true if it's possible to update table GSIs.
+// we can only perform one GSI create/update/delete at once.
+func canUpdateTableGSIs(r *resource) bool {
+	for _, gsiDescription := range r.ko.Status.GlobalSecondaryIndexesDescriptions {
+		if *gsiDescription.IndexStatus != svcsdk.IndexStatusActive {
+			return false
+		}
+	}
+	return true
+}
+
+// computeGlobalSecondaryIndexDelta compares two GlobalSecondaryIndex arrays and
+// return three different list containing the added, updated and removed
+// GlobalSecondaryIndex. The removed array only contains the IndexName of the
+// GlobalSecondaryIndex.
+func computeGlobalSecondaryIndexDelta(
+	a []*v1alpha1.GlobalSecondaryIndex,
+	b []*v1alpha1.GlobalSecondaryIndex,
+) (added, updated []*v1alpha1.GlobalSecondaryIndex, removed []string) {
+	var visitedIndexes []string
+loopA:
+	for _, aElement := range a {
+		visitedIndexes = append(visitedIndexes, *aElement.IndexName)
+		for _, bElement := range b {
+			if *aElement.IndexName == *bElement.IndexName {
+				if !equalGlobalSecondaryIndexes(aElement, bElement) {
+					updated = append(updated, bElement)
+				}
+				continue loopA
+			}
+		}
+		removed = append(removed, *aElement.IndexName)
+
+	}
+	for _, bElement := range b {
+		if !ackutil.InStrings(*bElement.IndexName, visitedIndexes) {
+			added = append(added, bElement)
+		}
+	}
+	return added, updated, removed
+}
+
+// equalGlobalSecondaryIndexesArrays returns true if two GlobalSecondaryIndex
+// arrays are equal regardless of the order of their elements.
+func equalGlobalSecondaryIndexesArrays(
+	a []*v1alpha1.GlobalSecondaryIndex,
+	b []*v1alpha1.GlobalSecondaryIndex,
+) bool {
+	added, updated, removed := computeGlobalSecondaryIndexDelta(a, b)
+	return len(added) == 0 && len(updated) == 0 && len(removed) == 0
+}
+
```

Comment:
I don't see this function being used anywhere?

Reply:
It is used in `customerPreCompare`:
```go
	if len(a.ko.Spec.GlobalSecondaryIndexes) != len(b.ko.Spec.GlobalSecondaryIndexes) {
		delta.Add("Spec.GlobalSecondaryIndexes", a.ko.Spec.GlobalSecondaryIndexes, b.ko.Spec.GlobalSecondaryIndexes)
	} else if a.ko.Spec.GlobalSecondaryIndexes != nil && b.ko.Spec.GlobalSecondaryIndexes != nil {
		if !equalGlobalSecondaryIndexesArrays(a.ko.Spec.GlobalSecondaryIndexes, b.ko.Spec.GlobalSecondaryIndexes) {
			delta.Add("Spec.GlobalSecondaryIndexes", a.ko.Spec.GlobalSecondaryIndexes, b.ko.Spec.GlobalSecondaryIndexes)
		}
	}
```